contracts/hydro/src/contract.rs

```diff
@@ -1497,6 +1551,118 @@ pub fn remove_liquidity_deployment(
+
+    let tokens_ratio_changes = vec![TokenGroupRatioChange {
+        token_group_id: token_group_id.clone(),
+        old_ratio,
+        new_ratio,
+    }];
+
+    apply_token_groups_ratio_changes(
+        deps.storage,
+        env.block.height,
+        current_round_id,
+        &tokens_ratio_changes,
+    )?;
+
+    let response = Response::new()
+        .add_attribute("action", "update_token_group_ratio")
+        .add_attribute("sender", info.sender)
+        .add_attribute("current_round_id", current_round_id.to_string())
+        .add_attribute("token_group_id", token_group_id.clone())
+        .add_attribute("old_ratio", old_ratio.to_string())
+        .add_attribute("new_ratio", new_ratio.to_string());
+
+    Ok(response)
+}
+
+pub fn add_token_info_provider(
+    mut deps: DepsMut<NeutronQuery>,
+    env: Env,
+    info: MessageInfo,
+    provider_info: TokenInfoProviderInstantiateMsg,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    let constants = load_current_constants(&deps.as_ref(), &env)?;
+
+    validate_contract_is_not_paused(&constants)?;
+    validate_sender_is_whitelist_admin(&deps, &info)?;
+
+    let token_info_provider_init_msgs =
+        add_token_info_providers(&mut deps, vec![provider_info.clone()])?;
+
+    let response = Response::new()
+        .add_attribute("action", "add_token_info_provider")
+        .add_attribute("sender", info.sender)
+        .add_attribute("token_info_provider", provider_info.to_string())
+        .add_submessages(token_info_provider_init_msgs);
+
+    Ok(response)
+}
+
+pub fn remove_token_info_provider(
+    deps: DepsMut<NeutronQuery>,
+    env: Env,
+    info: MessageInfo,
+    provider_id: String,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    let constants = load_current_constants(&deps.as_ref(), &env)?;
+
+    validate_contract_is_not_paused(&constants)?;
+    validate_sender_is_whitelist_admin(&deps, &info)?;
+
+    let mut token_info_provider =
+        match TOKEN_INFO_PROVIDERS.may_load(deps.storage, provider_id.clone())? {
+            Some(provider) => provider,
+            None => {
+                return Err(new_generic_error(format!(
+                    "Token info provider with ID: {} doesn't exist.",
+                    provider_id.clone()
+                )))
+            }
+        };
+
+    let current_round_id = compute_current_round_id(&env, &constants)?;
+
+    let tokens_ratio_changes: Vec<TokenGroupRatioChange> = token_info_provider
+        .get_all_token_group_ratios(&deps.as_ref(), current_round_id)?
+        .iter()
+        .map(|token_group| TokenGroupRatioChange {
+            token_group_id: token_group.0.clone(),
+            old_ratio: *token_group.1,
+            new_ratio: Decimal::zero(),
+        })
+        .collect();
+
+    // Remove any voting power on proposals and rounds that comes from tokens of the given token info provider.
+    apply_token_groups_ratio_changes(
```

Comment:
If we are removing LSM token info provider, this can be very gas consuming and potentially too large tx to enter the block. The only mitigation that we have at this moment is to perform the removal at the beginning of the round when there are no proposals. Updating just the rounds total power to remove all validators token groups should work.

One way to handle this properly would be to only store the token groups that need to be removed and introduce a new permission-less action to remove them in chunks. Since we are unlikely to remove LSM token info provider in the near future, I wouldn't do this part just yet.

Reply:
Thanks, good idea. Neutron is about to raise gas ~10x with their Mercury upgrade, so maybe they will make our life much easier :D